### PR TITLE
MPA-72_get-policies

### PR DESCRIPTION
**Pull request by andresantos252**
This PR includes the code for retrieving event policies from the API for a certain user.
The following diagram shows how the feature was implemented:

GetPolicies

Additionally, the Ktor plugin was used to perform API requests, serialisation of the data classes and also, when testing, to mock the API.

Finally, both unit tests and integration tests are included.

### CODE REVIEW

File: shared/src/commonMain/kotlin/com/mindera/minderapeople/apiclient/ApiHttpClient.kt [hunk 1]
```diff
@@ -0,0 +1,22 @@
+package com.mindera.minderapeople.apiclient
+
+import io.ktor.client.*
+import io.ktor.client.engine.*
+import io.ktor.client.plugins.contentnegotiation.*
+import io.ktor.serialization.kotlinx.json.*
+import kotlinx.serialization.json.Json
+
+
+
```
Comment (beloso) [inline shared/src/commonMain/kotlin/com/mindera/minderapeople/apiclient/ApiHttpClient.kt:10]: One white line is enough

File: shared/src/commonTest/kotlin/com/mindera/minderapeople/unit/apiclient/PolicyApiClientTest.kt [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+import io.ktor.client.engine.mock.*
+import io.ktor.http.*
+import io.ktor.utils.io.*
+import kotlinx.coroutines.runBlocking
+import kotlinx.serialization.json.Json
+import kotlin.test.Test
+import kotlin.test.assertEquals
+import kotlin.test.assertTrue
+
+
```
Comment (beloso) [inline shared/src/commonTest/kotlin/com/mindera/minderapeople/unit/apiclient/PolicyApiClientTest.kt:15]: one white line is enough

File: shared/src/commonMain/kotlin/com/mindera/minderapeople/repository/PolicyRepository.kt [hunk 3]
```diff
@@ -1,6 +1,13 @@
...
 
-import com.mindera.minderapeople.MinderaPeopleAPIClient
+import com.mindera.minderapeople.dto.PolicyDTO
+import com.mindera.minderapeople.apiclient.interfaces.IPolicyApiClient
+import com.mindera.minderapeople.repository.interfaces.IPolicyRepository
 
-class PolicyRepository(val apiClient: MinderaPeopleAPIClient) {
+class PolicyRepository(private val apiClient: IPolicyApiClient): IPolicyRepository {
+    override suspend fun getPolicies(userId: String): Result<List<PolicyDTO>> {
+        if(userId.matches(Regex("........-....-....-....-............")))
```
Comment (beloso) [inline shared/src/commonMain/kotlin/com/mindera/minderapeople/repository/PolicyRepository.kt:9]: is this the best Regex for UUID? or GUID? (What ID type is this?)

Comment (andresantos252) [reply shared/src/commonMain/kotlin/com/mindera/minderapeople/repository/PolicyRepository.kt]: userid verification removed, its not necessary here

**Review by asos-tiagoveloso — APPROVED** [review]

**Review by beloso — APPROVED** [review]

File: shared/src/commonMain/kotlin/com/mindera/minderapeople/apiclient/PolicyApiClient.kt [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+
+class PolicyApiClient(engine: HttpClientEngine): IPolicyApiClient {
+
+    private val apiHttpClient = ApiHttpClient(engine)
+    private val httpClient = apiHttpClient.httpClient
+
+    override suspend fun getAllPolicies(userId: String): Result<List<PolicyDTO>> {
+        return try {
+            val response = httpClient.get("${apiHttpClient.url}/policies") {
+                parameter("userId", userId)
```
Comment (JoaoRaimundo) [inline shared/src/commonMain/kotlin/com/mindera/minderapeople/apiclient/PolicyApiClient.kt:18]: If this parameter is reused within the api, consider having it as protected or internal.

File: shared/src/commonTest/kotlin/com/mindera/minderapeople/integration/policy/PolicyIntegrationTests.kt [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+                content = ByteReadChannel(Json.encodeToString(DefaultTestData.SUCCESSFUL_3_POLICIES)),
+                status = HttpStatusCode.OK,
+                headers = headersOf(HttpHeaders.ContentType, "application/json")
+            )
+        }
+        val apiClient = PolicyApiClient(mockEngine)
+        val repo = PolicyRepository(apiClient)
+
+        runBlocking {
+            val result = repo.getPolicies("a74b7fef-6c57-49c6-8c7c-2522a4defc70")
```
Comment (JoaoRaimundo) [inline shared/src/commonTest/kotlin/com/mindera/minderapeople/integration/policy/PolicyIntegrationTests.kt:31]: do we need this type of key here? In tests, a mock key should be enough, unless we are testing a key validator, that shouldn't be the case.

File: shared/src/commonMain/kotlin/com/mindera/minderapeople/apiclient/ApiHttpClient.kt [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+class ApiHttpClient(engine: HttpClientEngine) {
+
+    val url = "http://localhost:3000/api"
+
+    val httpClient = HttpClient(engine) {
+        install(ContentNegotiation) {
+            json(Json {
+                ignoreUnknownKeys = true
+                useAlternativeNames = false
```
[the remark this answers is not in the document]
Comment (andresantos252) [reply shared/src/commonMain/kotlin/com/mindera/minderapeople/apiclient/ApiHttpClient.kt:17]: Fixed in PR #3

File: shared/src/commonTest/kotlin/com/mindera/minderapeople/integration/policy/PolicyIntegrationTests.kt (hunk 5 above)
Comment (andresantos252) [reply shared/src/commonTest/kotlin/com/mindera/minderapeople/integration/policy/PolicyIntegrationTests.kt]: Fixed

**Pull request merged**